qiskit_dynamics/solvers/fixed_step_solvers.py

```diff
@@ -100,7 +109,83 @@ def take_step(generator, t0, y, h):
         eval_time = t0 + (h / 2)
         return expm(generator(eval_time) * h) @ y
 
+    # ensure the output of rhs_func is a raw array
+    def wrapped_rhs_func(*args):
+        return Array(generator(*args)).data
+
     return fixed_step_solver_template(
+        take_step, rhs_func=wrapped_rhs_func, t_span=t_span, y0=y0, max_dt=max_dt, t_eval=t_eval
+    )
+
+
+def lanczos_diag_solver(
+    generator: Callable,
+    t_span: Array,
+    y0: Array,
+    max_dt: float,
+    t_eval: Optional[Union[Tuple, List, Array]] = None,
+    k_dim: Optional[int] = None,
+):
+    """Fixed-step size matrix exponential based solver implemented using
+    lanczos algorithm. Solves the specified problem by taking steps of
+    size no larger than ``max_dt``.
+
+    Args:
+        generator: Generator for the LMDE.
+        t_span: Interval to solve over.
+        y0: Initial state.
+        max_dt: Maximum step size.
+        t_eval: Optional list of time points at which to return the solution.
+        k_dim: Integer which specifies the dimension of Krylov subspace used for
+               lanczos iteration. Acts as an accuracy parameter. ``k_dim`` must
+               always be less than or equal to dimension of generator.
+
+    Returns:
+        OdeResult: Results object.
+
+    Raises:
+        QiskitError: If ``k_dim`` is greater than dimension of generator.
+    """
+
+    dim = generator(0).shape[0]
```

Comment:
This should be `generator(t_span[0])`, not `generator(0)`.

Reply:
Good catch! Thanks.